vowpalwabbit/core/include/vw/core/learner.h

```diff
@@ -297,386 +160,175 @@ class learner
-      const base_learner* l2, VW::workspace& output_ws, base_learner* output_l)
-  {
-    auto name = output_l->get_name();
-    assert(name == l1->get_name());
-    assert(name == l2->get_name());
-    if (_subtract_with_all_fn != nullptr)
-    {
-      _subtract_with_all_fn(
-          ws1, l1->_learner_data.get(), ws2, l2->_learner_data.get(), output_ws, output_l->_learner_data.get());
-    }
-    else if (_subtract_fn != nullptr)
-    {
-      _subtract_fn(l1->_learner_data.get(), l2->_learner_data.get(), output_l->_learner_data.get());
-    }
-    else { THROW("learner " << name << " does not support subtraction to generate a delta."); }
-  }
-
-  VW_ATTR(nodiscard) bool has_legacy_finish() const { return _finish_example_fd.finish_example_f != nullptr; }
-  VW_ATTR(nodiscard) bool has_update_stats() const { return _finish_example_fd.update_stats_f != nullptr; }
-  VW_ATTR(nodiscard) bool has_print_update() const { return _finish_example_fd.print_update_f != nullptr; }
-  VW_ATTR(nodiscard) bool has_output_example_prediction() const
-  {
-    return _finish_example_fd.output_example_prediction_f != nullptr;
-  }
-  VW_ATTR(nodiscard) bool has_cleanup_example() const { return _finish_example_fd.cleanup_example_f != nullptr; }
-  VW_ATTR(nodiscard) bool has_merge() const { return (_merge_with_all_fn != nullptr) || (_merge_fn != nullptr); }
-  VW_ATTR(nodiscard) bool has_add() const { return (_add_with_all_fn != nullptr) || (_add_fn != nullptr); }
-  VW_ATTR(nodiscard) bool has_subtract() const
-  {
-    return (_subtract_with_all_fn != nullptr) || (_subtract_fn != nullptr);
-  }
+  // Get prediction and label types
   VW_ATTR(nodiscard) prediction_type_t get_output_prediction_type() const { return _output_pred_type; }
   VW_ATTR(nodiscard) prediction_type_t get_input_prediction_type() const { return _input_pred_type; }
   VW_ATTR(nodiscard) label_type_t get_output_label_type() const { return _output_label_type; }
   VW_ATTR(nodiscard) label_type_t get_input_label_type() const { return _input_label_type; }
+
+  // Is this learner multiline or singleline?
   VW_ATTR(nodiscard) bool is_multiline() const { return _is_multiline; }
+
+  // Get the reduction's name
   VW_ATTR(nodiscard) const std::string& get_name() const { return _name; }
-  VW_ATTR(nodiscard) const base_learner* get_learn_base() const { return _learn_fd.base; }
-  VW_ATTR(nodiscard) base_learner* get_learn_base() { return _learn_fd.base; }
-  /// If true, this specific learner defines a save load function. If false, it simply forwards to a base
-  /// implementation.
-  VW_ATTR(nodiscard) bool learner_defines_own_save_load() { return _learn_fd.data == _save_load_fd.data; }
+
+  // Returns a pointer to the previous learner in the reduction stack
+  // Returns nullptr if this is the base learner of the stack
+  VW_ATTR(nodiscard) const learner* get_previous_learner() const { return _previous_learner.get(); }
```

Comment:
Let's revisit this rename, previous is not the right word here. Status quo is that a reduction only knows whats under it i.e. a generic promise of a base learner - which might have some extra layers on top but that is hidden away.

Reply:
Maybe it's helpful to separate the issues when thinking about naming things. There are 2 tensions:

From the stack construction perspective, we move in one direction (bottom of the stack to top). When predict() is called, we move in the other direction (top of stack calls the next reduction down the stack).
The word 'base' has two meanings when going down the predict/learn stack. (a) It's the next down stream reduction. (b) It's the last reduction in the stack.

Reply:
The stack gets built from the top down also not from the bottom to the top. https://github.com/VowpalWabbit/vowpal_wabbit/blob/2fe6f2fb97962040c674e1bee3665d086347f2a0/vowpalwabbit/core/src/reduction_stack.cc#L273

We pop and call setup func - but yes behind the scenes some internal impl detail is that the learner instance for the bottom setup function gets called first because to build a stack you need the previous element first before wrapping it up with the next but that's just normal recursion/stack side effect of how things get executed.

Reply:
Yeah I agree with that. It is still built top down, it is just recursive in how it calls.

Reply:
I have now renamed base learner (bottom-most learner of the stack) to "foundation learner". It turns out that this is the option that minimizes the number of other changes that needed to be made, especially to public documentation.

Changes to documentation are mainly in this wiki page, for which I have created a new copy: https://github.com/VowpalWabbit/vowpal_wabbit/wiki/What-is-a-learner%3F-(new-version)

In addition, this page uses "base" in both ways, but mostly in the "base of a learner" sense. It will require one small change from the other meaning of base to "foundation". https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Matching-Label-and-Prediction-Types-Between-Reductions

There are a few other wiki pages where base would get changed to foundation, but those are fairly minor changes.

I could not find any change needed on the VW site, but maybe I missed something. https://vowpalwabbit.org/docs/vowpal_wabbit/python/latest/index.html

In summary, we have 3 options
1. Rename the bottom-most learner to "foundation". This is the current state of this PR, and will require the changes described above to make our documentation consistent.
2. Rename a reduction's base to "previous". This was what I has first done, but it was ambiguous because the next learner that an example would travel to was named "previous". We could also name it "next", but that gives the counterintuitive situation where a reduction is created by copying its "next" learner, which already exists before itself is created.
3. Don't rename anything, and leave the double meaning of "base" as-is.

Reply:
Also tagging @JohnLangford for any comments on this